File: Algorithm_304_Range Sum Query 2D - Immutable.cpp
#include <iostream>
#include <vector>
using namespace std;
class NumMatrix
{
private:
    vector<vector<int>> mySumMatrix;
public:
    NumMatrix(vector<vector<int>> matrix)
    {
        int rowLen = matrix.size();
        int colLen = rowLen > 0 ? matrix[0].size() : 0;
        vector<int> tmp(colLen + 1, 0);
        for (int i = 0; i < rowLen + 1; i++) mySumMatrix.push_back(tmp);
        for (int i = 0; i < rowLen; i++) for (int j = 0; j < colLen; j++) mySumMatrix[i + 1][j + 1] = mySumMatrix[i][j + 1] + mySumMatrix[i + 1][j] - mySumMatrix[i][j] + matrix[i][j];
    }

    int sumRegion(int row1, int col1, int row2, int col2)
    {
        return mySumMatrix[row2 + 1][col2 + 1] + mySumMatrix[row1][col1] - mySumMatrix[row1][col2 + 1] - mySumMatrix[row2 + 1][col1];
    }
};

File: Algorithm_304_Range Sum Query 2D - Immutable.java
// 思路：把以matrix[0][0]为左上角、matrix[i][j]为右下角的矩形中所有元素和存放在sums[i + 1][j + 1]中
// 则计算sumRegion时，只需要把以matrix[row2][col2]为右下角的矩形中所有元素之和
// 减去以matrix[row2][col1 - 1]为右下角的矩形中所有元素之和
// 再减去以matrix[row1 - 1][col2]为右下角的矩形中的所有元素之和
// 再加上以matrix[row1 - 1][col1 - 1]为右下角的矩形中的所有元素之和
// 即为答案

class NumMatrix {
    private int[][] sums;
    public NumMatrix(int[][] matrix) {
        int m = matrix.length, n = matrix.length == 0 ? 0 : matrix[0].length;
        sums = new int[m + 1][n + 1];
        for (int i = 1; i <= matrix.length; i++) {
            for (int j = 1; j <= matrix[0].length; j++) {
                sums[i][j] = matrix[i - 1][j - 1] + sums[i - 1][j] + sums[i][j - 1] - sums[i - 1][j - 1];
            }
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        return sums[row2 + 1][col2 + 1] - sums[row2 + 1][col1] - sums[row1][col2 + 1] + sums[row1][col1];
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */